dipy/viz/ui.py

```diff
@@ -28,16 +28,13 @@ class UI(object):
 
     Attributes
     ----------
-    ui_param : object
-        This is an attribute that can be passed to the UI object by the
-        interactor.
-    ui_list : list of :class:`UI`
-        This is used when there are more than one UI elements inside
-        a UI element. They're all automatically added to the renderer at the
-        same time as this one.
-    parent_ui: UI
-        Reference to the parent UI element. This is useful of there is a parent
-        UI element and its reference needs to be passed down to the child.
```

Comment:
Looking at your Checkbox branch, wouldn't it be that making the `key_press` method a member of the `Checkbox` class rather than a staticmethod solves the problem? You would be able to use `self.` instead of `textbox_obj.parent_ui`.

That said, it is clear that the old way of dealing with events is broken. For instance, your Checkbox UI component needed to capture the `on_key_press` event and essentially re-implement the textbox behavior (i.e. the same code is used in the TextBox and Checkbox component).

Also, why does the Checkbox need a TextBox rather than a TextBlock (i.e. label)?

Reply:
Also, should we make the a distinction between Checkbox and GroupOfCheckboxes?

Reply:
What I meant is we should have an atomic Checkbox UI component. Like
```python
class Checkbox(UI):

    def __init__(self, label):
        """
        Parameters
        ----------
        label : str
            Checkbox's label.

        """
        super(Checkbox, self).__init__()
        self.text.message = label

    def _setup(self):
        """ Setup this UI component.

        TODO
        """
        # Checkbox's button
        self.button_icons = {}
        self.button_icons['unchecked'] = read_viz_icons(fname="stop2.png")
        self.button_icons['checked'] = read_viz_icons(fname="checkmark.png")
        self.button = Button2D(icon_fnames=self.button_icons)

        self.text = TextBlock2D()

        # Add default events listener for this UI component.
        self.button.on_left_mouse_button_pressed = self.toggle_check

    def _get_actors(self):
        """ Get the actors composing this UI component.
        """
        return self.button.actors + self.text.actors

    def _add_to_renderer(self, ren):
        """ Add all subcomponents or VTK props that compose this UI component.

        Parameters
        ----------
        ren : renderer
        """
        self.button.add_to_renderer(ren)
        self.text.add_to_renderer(ren)

    def _get_size(self):
        # Consider the handle's size when computing the slider's size.
        width = self.button.size[0] + 5 + self.text.size[0]
        height = max(self.button.size[1], self.text.size[1])
        return np.array([width, height])

    def _set_position(self, coords):
        """ Position the lower-left corner of this UI component.

        Parameters
        ----------
        coords: (float, float)
            Absolute pixel coordinates (x, y).
        """
        # [x] Label
        self.button.position = coords
        offset = (self.button.size[0] + 5, 0)
        self.text.position = self.button.position + offset

    def toggle_check(self, i_ren, obj, button_object):
        self.button.next_icon()
        i_ren.force_render()
```
Then, have another UI Component that have a group of Checkboxes if desired which should create instances of SingleCheckbox and add them to a Panel2D.

Reply:
Yeah this is pretty good.

But for RadioButton we can't have a `toggle_check` for each option. That we will have to add in the GroupOfRadioButton class. So maybe, I can make an Option class like yours without the `toggle_check` function and use this class for both GroupOfCheckbox and GroupOfRadioButton

Reply:
A radio button component could be seen as a "GroupOfCheckBox" with some restrictions on how many "Checkbox" can be selected. Also, each Checkbox should have a "on_change" callback function that gets triggered whenever a Checkbox changes state (similar to https://github.com/nipy/dipy/pull/1492/files#diff-fa451eb394ebc769b7b376b567c756b5R149)